op-plasma/damgr.go

```diff
@@ -2,48 +2,420 @@ package plasma
+	// chain so potential future challenge events can be selected.
+	ch := d.state.GetOrTrackChallenge(comm.Encode(), blockId.Number, d.cfg.ChallengeWindow)
+
+	// Fetch the input from the DA storage.
+	data, err := d.storage.GetInput(ctx, comm)
+
+	// data is not found in storage but may be available if the challenge was resolved.
+	notFound := errors.Is(ErrNotFound, err)
+
+	if err != nil && !notFound {
+		d.log.Error("failed to get preimage", "err", err)
+		// the storage client request failed for some other reason
+		// in which case derivation pipeline should be retried
+		return nil, err
+	}
+
+	switch ch.challengeStatus {
+	case ChallengeActive:
+		if d.isExpired(ch.expiresAt) {
+			// this challenge has expired, this input must be skipped
+			return nil, ErrExpiredChallenge
+		} else if notFound {
+			// data is missing and a challenge is active, we must wait for the challenge to resolve
+			// hence we continue syncing new origins to sync the new challenge events.
+			if err := d.LookAhead(ctx); err != nil {
+				return nil, err
+			}
+			return nil, ErrPendingChallenge
+		}
+	case ChallengeExpired:
+		// challenge was marked as expired, skip
+		return nil, ErrExpiredChallenge
+	case ChallengeResolved:
+		// challenge was resolved, data is available in storage, return directly
+		if !notFound {
+			return data, nil
+		}
+		// data not found in storage, return from challenge resolved input
+		resolvedInput, err := d.state.GetResolvedInput(comm.Encode())
+		if err != nil {
+			return nil, err
+		}
+		return resolvedInput, nil
+	default:
+		if notFound {
+			if d.isExpired(ch.expiresAt) {
+				// we're past the challenge window and the data is not available
+				return nil, ErrMissingPastWindow
+			} else {
+				// continue syncing challenges hoping it eventually is challenged and resolved
+				if err := d.LookAhead(ctx); err != nil {
+					return nil, err
+				}
+				return nil, ErrPendingChallenge
+			}
+		}
+	}
+
+	return data, nil
+}
+
+// isExpired returns whether the given expiration block number is lower or equal to the current head
+func (d *DA) isExpired(bn uint64) bool {
+	return d.origin.Number >= bn
+}
+
+// AdvanceL1Origin syncs any challenge events included in the l1 block, expires any active challenges
+// after the new resolveWindow, computes and signals the new finalized head and sets the l1 block
+// as the new head for tracking challenges. If forwards an error if any new challenge have expired to
+// trigger a derivation reset.
+func (d *DA) AdvanceL1Origin(ctx context.Context, block eth.BlockID) error {
+	// do not repeat for the same origin
+	if block.Number <= d.origin.Number {
+		return nil
+	}
+	// sync challenges for the given block ID
+	if err := d.LoadChallengeEvents(ctx, block); err != nil {
+		return err
+	}
+	// advance challenge window, computing the finalized head
+	bn, err := d.state.ExpireChallenges(block.Number)
+	if err != nil {
+		// warn the reset function not to clear the state
+		d.resetting++
```

Comment:
In theory we could expire multiple challenges in one L1 block which would seem to throw off the reset count.